EmpowerPlant/ShoppingCart.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  ShoppingCart.swift
+//  EmpowerPlant
+//
+//  Created by William Capozzoli on 4/12/22.
+//
+
+import Foundation
+
+
+class ShoppingCart {
+    
+    //creates the instance and guarantees that it's unique
+    static let instance = ShoppingCart()
+    
+    private init() {
+    }
+    
+    var items = [Product]()
+    var total = 0
+    var quantities = Quantities()
+    
+    // This updates the items, total, and quantities
+    static func addProduct(product: Product) {
+        
+        if !self.instance.items.contains(product) {
+            self.instance.items.append(product)
+        }
+        
+        let productId = product.productId!
+        let id = Int(productId)
+        
+        // These are the values as used in the backend database in CloudSQL
+        /*
+         Plant Mood 3
+         Botana Voice 4
+         Plant Stroller 5
+         Plant Nodes 6
+        */
+        switch id {
+        case 3:
```

Comment:
(note: this is not a blocker for approving this PR, just a thought)

The repetition of these kinds of sections several times seems not-ideal:

```
self.instance.quantities.plantMood += 1
let price = Int(product.price!)
self.instance.total = self.instance.total + price!
print("> quantities plantMood", self.instance.quantities.plantMood)
break
```

I was originally thinking we could clean this up and not repeat code. Maybe dynamically call `self.instance.quantities.<the product name>`, i.e. call a function based on a string. I didn't know the 'Swift way' to do this, but when I googled it seemed like it's actively discouraged from doing that. They offer another idea, but I'm not sure that really buys us much 🤷  https://www.hackingwithswift.com/forums/swift/calling-object-methods-dynamically/1983

All this is to say that I think this section could possibly be cleaned up, but I don't know the right way of doing it without doing a bunch of additional research (which is not worth it because this is straightforward/contained enough that it doesn't actually create a problem), so I have nothing valuable to add and am fine to move forward with what's here. 😄

Reply:
@cstavitsky 
I can try `self.instance.quantities.<the product name>` but the problem is that's a registered property at compile-time. Swift doesn't allow things like self.instance.quantities["propertyName"]. I know there's a getter for this and as long as it existed at compile-time, it may access it. But that's only for 'getting' and not 'setting'. I tried for an hour to build those ["propertyNames"]'s on the fly, so it'd make 4 if it's 4 products or 1,000 if it's 1,000 products but couldn't get it to work. Here's an [explanation about the problem in swift](https://stackoverflow.com/questions/26667380/in-swift-for-anyobject-how-do-i-setvalue-then-call-valueforkey). I could try changing the data structure altogether, but then I'd have to later on transform to match the backend's data structure. Better to keep it the same as what we have in JS, python, java, etc.

The real data structure (coming from the Empower Plant in javascript) has numbers as keys, for the quantity, and that's not allowed in swift, like it is in Javascript. So this is why I ended up with this weird data structure `self.instance.quantities.<the product name>` in ShoppingCart.swift, and the two switch statements you saw, which are accomplishing different things. I had to convert that dictionary to an array for the shopping cart UI, because the Table works off of an array (and not a dictionary, which is what Empower Plant in javascript uses).

I looked at your link and it says "Swift doesn't really do dynamism like that." yeah it's a lot of code to accomplish something simple. It was hard to follow. Switch statements and declaring the products as properties works as it's only 4 products and not 10,000 products. No plans for +4 products any time soon (creative team created those).

Yeah I can put a helper function in there for that switch statement.

Happy to present the weirdness about it on a call. It's definitely not ideal but works for now.

Reply:
I'm trying to make
```
        case 6:
            updateQuantitiesAndTotal(self.instance.quantities.plantNodes, product)
            break
```

with
```
    static func updateQuantitiesAndTotal(productQuantity: <Type>, product: product) {
       // this doesn't work, and I can't find a way to do it with Generics in Swift yet
       self.instance.quantities[<Type>] += 1
            let price = Int(product.price!)
            self.instance.total = self.instance.total + price!
        print("> updateQuantitiesAndTotal")
    }
```
Also tried
```
    static func updateQuantitiesAndTotal(productName: String, product: Product) {
       // this doesn't compile, hence I thought about using Generic Types
        self.instance.quantities["productName"] += 1
            let price = Int(product.price!)
            self.instance.total = self.instance.total + price!
        print("> updateQuantitiesAndTotal")
    }
```


Ok so found a way to separate out the updateTotal at least. That's what you'll see in the new code.


I tried to do updateQuantitiesAndTotal but ran into the same problems as before, and the code started getting hairy, losing readability, and needing yet more switch statements and helpers
<img width="1102" alt="image" src="https://user-images.githubusercontent.com/8920574/163567795-c93bb75b-9ce8-46c9-9250-55c1a5ea67eb.png">